### PR TITLE
wal: use page buffered writer for writing records

### LINKED ISSUE #6271: read wal error (walpb: crc mismatch) and cannot be repaired
Happened in *failpoints* test.

```
2016-08-26 00:51:25.427398 I | etcd-tester: [round#29 case#43] injecting failure "failpoint github.com/coreos/etcd/mvcc/backend/beforeCommit panic one"
2016-08-26 00:51:25.474207 I | etcd-tester: transport: http2Client.notifyError got notified that the client transport was broken EOF.
2016-08-26 00:51:25.474925 I | etcd-tester: grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 10.240.0.22:2379: getsockopt: connection refused"; Reconnecting to {"10.240.0.22:2379" <nil>}
...
2016-08-26 00:57:28.222375 W | etcd-tester: #59 setHealthKey error (grpc: timed out when dialing (http://10.240.0.22:2379))
2016-08-26 00:57:28.222437 I | etcd-tester: Failed to dial 10.240.0.22:2379: context canceled; please retry.
2016-08-26 00:57:29.222529 I | etcd-tester: [round#29 case#43] recovery error: grpc: timed out when dialing (http://10.240.0.22:2379)
2016-08-26 00:57:29.222607 I | etcd-tester: [round#29 case#43] cleaning up...
```

```
# in 10.240.0.22
2016-08-26 00:51:28.445205 I | etcdmain: etcd Version: 3.0.0+git
2016-08-26 00:51:28.445252 I | etcdmain: Git SHA: f1ead43-FAILPOINTS
2016-08-26 00:51:28.445257 I | etcdmain: Go Version: go1.7
2016-08-26 00:51:28.445262 I | etcdmain: Go OS/Arch: linux/amd64
2016-08-26 00:51:28.445267 I | etcdmain: setting maximum number of CPUs to 4, total number of available CPUs is 4
2016-08-26 00:51:28.445303 N | etcdmain: the server is already initialized as member before, starting as etcd member...
2016-08-26 00:51:28.445312 I | etcdmain: advertising using detected default host "10.240.0.22"
2016-08-26 00:51:28.445339 I | embed: listening for peers on http://10.240.0.22:2380
2016-08-26 00:51:28.445366 I | embed: listening for client requests on 10.240.0.22:2379
2016-08-26 00:51:28.465383 I | etcdserver: recovered store from snapshot at index 15609237
2016-08-26 00:51:28.465414 I | etcdserver: name = etcd-2
2016-08-26 00:51:28.465420 I | etcdserver: data dir = agent.etcd
2016-08-26 00:51:28.465426 I | etcdserver: member dir = agent.etcd/member
2016-08-26 00:51:28.465432 I | etcdserver: heartbeat = 100ms
2016-08-26 00:51:28.465437 I | etcdserver: election = 1000ms
2016-08-26 00:51:28.465441 I | etcdserver: snapshot count = 10000
2016-08-26 00:51:28.465452 I | etcdserver: advertise client URLs = http://10.240.0.22:2379
2016-08-26 00:51:28.752071 C | etcdserver: read wal error (walpb: crc mismatch) and cannot be repaired
```

It was deployed with @heyitsanthony 's branch https://github.com/coreos/etcd/pull/6267.

WAL file read shows

```
000090 00 00 00 00 85 08 02 10 a6 dd a3 df 04 1a 98 01
0000a0 08 00 10 be 09 18 9a 88 b9 07 22 8b 01 22 7b 0a
0000b0 13 66 6f 6f 30 00 00 00 00 00 00 00 00 00 00 00
0000c0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
*
d58610 00 00 00 00 00
d58615
```

### PR DESCRIPTION
**Pull request by heyitsanthony**
Fixes #6271
 
/cc @gyuho

### CODE REVIEW

**Comment by gyuho** [comment]
Ok I will try this branch in failpoints cluster.

File: wal/encoder.go [hunk 1]
```diff
@@ -36,7 +36,7 @@ type encoder struct {
 
 func newEncoder(w io.Writer, prevCrc uint32) *encoder {
 	return &encoder{
-		bw:  bufio.NewWriter(w),
+		bw:  ioutil.NewPageWriter(w, 4096),
```
Comment (xiang90) [inline wal/encoder.go]: define 4096 somewhere?

Comment (gyuho) [inline wal/encoder.go]: Also doc that this is buffering same as bufio https://github.com/golang/go/blob/master/src/bufio/bufio.go#L18?

Comment (heyitsanthony) [inline wal/encoder.go]: @gyuho it's only a coincidence the wal page size is the bufio default buffer size; the full buffer size for a pagewriter is 128k. I'll add a comment about the choice of the value.

Comment (gyuho) [inline wal/encoder.go]: Got it. Thanks!

**Comment by xiang90** [comment]
LGTM. Have you tried to run the benchmark in wal pkg to see if there is any performance gain?

**Comment by heyitsanthony** [comment]
@xiang90 benchmark found a bug in the page writer 👍 

Master:
```
BenchmarkWrite100EntryWithoutBatch-4    	     200	  10144426 ns/op	   0.01 MB/s
BenchmarkWrite100EntryBatch10-4         	    2000	    943462 ns/op	   0.11 MB/s
BenchmarkWrite100EntryBatch100-4        	   10000	    116260 ns/op	   0.93 MB/s
BenchmarkWrite100EntryBatch500-4        	   50000	     28449 ns/op	   3.80 MB/s
BenchmarkWrite100EntryBatch1000-4       	  100000	     15095 ns/op	   7.15 MB/s
BenchmarkWrite1000EntryWithoutBatch-4   	     100	  10645949 ns/op	   0.09 MB/s
BenchmarkWrite1000EntryBatch10-4        	    2000	   1016150 ns/op	   0.99 MB/s
BenchmarkWrite1000EntryBatch100-4       	   10000	    144891 ns/op	   6.96 MB/s
BenchmarkWrite1000EntryBatch500-4       	   50000	     42404 ns/op	  23.79 MB/s
BenchmarkWrite1000EntryBatch1000-4      	   50000	     28502 ns/op	  35.40 MB/s
```

This patch:
```
BenchmarkWrite100EntryWithoutBatch-4    	     100	  10629690 ns/op	   0.01 MB/s
BenchmarkWrite100EntryBatch10-4         	    2000	    963754 ns/op	   0.11 MB/s
BenchmarkWrite100EntryBatch100-4        	   10000	    107374 ns/op	   1.01 MB/s
BenchmarkWrite100EntryBatch500-4        	  100000	     23026 ns/op	   4.69 MB/s
BenchmarkWrite100EntryBatch1000-4       	  200000	     11675 ns/op	   9.25 MB/s
BenchmarkWrite1000EntryWithoutBatch-4   	     100	  10485759 ns/op	   0.10 MB/s
BenchmarkWrite1000EntryBatch10-4        	    2000	    997684 ns/op	   1.01 MB/s
BenchmarkWrite1000EntryBatch100-4       	   10000	    121113 ns/op	   8.33 MB/s
BenchmarkWrite1000EntryBatch500-4       	   50000	     25324 ns/op	  39.84 MB/s
BenchmarkWrite1000EntryBatch1000-4      	  100000	     13901 ns/op	  72.58 MB/s
```

**Comment by gyuho** [comment]
I can confirm that this fixes https://github.com/coreos/etcd/issues/6271 since it used to fail with less than 3 rounds. Now it's over 10 rounds. Thanks!

**Comment by xiang90** [comment]
lgtm

**Pull request merged**